### PR TITLE
Address #120, Implement incoherent/coherent signal subgraph estimators

### PR DESCRIPTION
**Pull request by gkang7**
#### Reference Issues/PRs

Addresses #120 
Refers to the algorithms outlines by https://arxiv.org/pdf/1108.1427.pdf

#### What does this implement/fix? Explain your changes.

This implements the incoherent and coherent algorithms from the graph classification paper, along with a function that handles contingency matrix construction. Includes tests for all of the additions.

Replication of Jovo's Fig 4 from the paper:
https://github.com/NeuroDataDesign/team-networkclass/blob/master/Gun/Summaries/figrep.png

#### Any other comments?

Tutorial notebook will be submitted later.

### CODE REVIEW

File: graspy/__init__.py [hunk 1]
```diff
@@ -9,6 +9,7 @@
 import graspy.plot
 import graspy.simulations
 import graspy.utils
+import graspy.subgraphing
```
Comment (j1c) [inline graspy/__init__.py]: Call this `subgraph` instead of `subgraphing`

File: graspy/subgraphing/sg.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    rowsum0 = len(labels) - rowsum1
+    for i in range(nverts):
+        for j in range(nverts):
+            a = sum(graphs[i, j, labels == 0])
+            b = sum(graphs[i, j, :]) - a
+            out[i, j, :, :] = [[a, rowsum0 - a], [b, rowsum1 - b]]
+    return out
+
+
+def estimate_signal_subgraph(graphs, labels, constraints):
```
Comment (j1c) [inline graspy/subgraphing/sg.py]: Turn this into a class instead of a function

Comment (gkang7) [reply graspy/subgraphing/sg.py]: Should I leave the contingency matrix function as is then?

Comment (j1c) [reply graspy/subgraphing/sg.py]: Yea. It can just be a private function in your class

Comment (gkang7) [reply graspy/subgraphing/sg.py]: Cool. I'll get started now

Comment (gkang7) [reply graspy/subgraphing/sg.py]: So at the moment, I only have the signal subgraph estimator and the contingency matrix constructor. Would it be okay if I just had

self.graphs and self.labels set through __init__ method
self.subgraph set through a transform method
And the contingency constructor through a private function, 
without having a normal fit method? Or is there a particular reason to have both a fit and fit_transform method?

Comment (gkang7) [reply graspy/subgraphing/sg.py]: Alternatively I can have the computation of the signal subgraph done through fit, and just return the actual subgraph through the transform method, which is what Ali did for faq

File: graspy/subgraph/sg.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    sigsub: tuple, shape (2, n_edges)
+        A tuple of a row index array and column index array, where n_edges is the size of the signal-subgraph determined by *constraints*.
+    
+    References
+    ----------
+    .. [1] J. T. Vogelstein, W. R. Gray, R. J. Vogelstein, and C. E. Priebe, "Graph Classification using Signal-Subgraphs: Applications in Statistical Connectomics," arXiv:1108.1427v2 [stat.AP], 2012.
+    
+    """
+
+    def __init__(self, graphs, labels):
```
Comment (j1c) [inline graspy/subgraph/sg.py]: I would have no inputs for class constructor here. graphs and labels should go to fit and fit_transform method

Comment (j1c) [reply graspy/subgraph/sg.py]: You can either abstract out the type checking to a separate private function, or just move it to fit

File: graspy/subgraph/sg.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        rowsum1 = sum(self.labels)
+        rowsum0 = len(self.labels) - rowsum1
+        for i in range(nverts):
+            for j in range(nverts):
+                a = sum(self.graphs[i, j, self.labels == 0])
+                b = sum(self.graphs[i, j, :]) - a
+                out[i, j, :, :] = [[a, rowsum0 - a], [b, rowsum1 - b]]
+        self.contmat = out
+
+    def fit(self, constraints):
```
Comment (j1c) [inline graspy/subgraph/sg.py]: When you put graphs and labels here as inputs, I would call them `X` and `y` for consistency with all other classes.

File: graspy/subgraph/sg.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+                    if wcounter > len(wset):
+                        sigsub = []
+                        wconv = 1
+        else:
+            msg = "Input constraints must be an int for the incoherent signal-subgraph estimator, or a vector of length 2 for the coherent subgraph estimator."
+            raise TypeError(msg)
+        self.sigsub = sigsub
+        return self
+
+    def fit_transform(self, constraints, return_mask=False):
```
Comment (j1c) [inline graspy/subgraph/sg.py]: Same thing. Add `X` and `y` as inputs here. I think you could just compute mask and save as class attribute called `mask_` and not having this option.

File: graspy/subgraph/sg.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    ----------
+    graphs: array-like, shape (n_vertices, n_vertices, s_samples)
+        A series of labeled (n_vertices, n_vertices) unweighted graph samples. If undirected, the upper or lower triangle matrices should be used.
+        
+    labels: vector, length (s_samples)
+        A vector of class labels. There must be a maximum of two classes.
+         
+    Attributes
+    ----------
+    contmat: array-like, shape (n_vertices, n_vertices, 2, 2)
```
Comment (j1c) [inline graspy/subgraph/sg.py]: computed attributes should end with `_`.

**Comment by j1c** [comment]
Nice. Very close to being merged. I just left some comments for minor changes.

File: graspy/subgraph/sg.py (hunk 3 above)
Comment (gkang7) [reply graspy/subgraph/sg.py]: So if I'm moving type checking to fit, and inputs to fit and fit_transform, is it convention to have an empty constructor? Or should I get rid of __init__ altogether
